### PR TITLE
TRINITY-52 Extend Announcements API

### PR DESCRIPTION
**Pull request by stetsche**
Added related commit from Adrian, which is on master, but not on the trinity branch yet.

I wanted to remove `getAnnouncementRange` from AnnouncementActions in favor of having it in the AnnouncementService, but found myself to deep in that rabbit hole.

### CODE REVIEW

**Review by profmikegreene — APPROVED** [review]
Looks good to me for what I need in Trinity/sui-table

File: announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/BaseAnnouncementService.java [hunk 1]
```diff
@@ -648,6 +650,30 @@ public Reference getAnnouncementReference(String context)
...
+			try {
+				Site site = m_siteService.getSite(m_entityManager.newReference(a.getReference()).getContext());
+				String allGroupString = a.getAnnouncementHeader().getGroups().stream().reduce("", (ourString, groupId) -> {
+					return ourString.concat(", " + site.getGroup(groupId).getTitle());
+				});
+				allGroupString = allGroupString.substring(2);
+				return allGroupString;
+			}
+			catch (IdUnusedException e) {
+				log.error("Site could not be loaded: {}", e.getMessage());
```
Comment (adrianfish) [inline announcement/announcement-impl/impl/src/java/org/sakaiproject/announcement/impl/BaseAnnouncementService.java:671]: I'd use e.toString() here. It has more info in it.

File: webapi/src/main/java/org/sakaiproject/webapi/beans/AnnouncementRestBean.java [hunk 2]
```diff
@@ -30,20 +38,40 @@
...
-        ResourceProperties props = am.getProperties();
-        String release = props.getProperty(AnnouncementService.RELEASE_DATE);
-        if (release != null) date = Long.parseLong(release);
-        this.url = url;
+        this.access = access;
+        AnnouncementMessageHeader header = am.getAnnouncementHeader();
+        subject = header.getSubject();
+        author = header.getFrom().getDisplayName();
+        date = header.getInstant().toEpochMilli();
+        hasAtachment = !header.getAttachments().isEmpty();
```
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/beans/AnnouncementRestBean.java:57]: typo there: hasAttachment

File: webapi/src/main/java/org/sakaiproject/webapi/beans/AnnouncementRestBean.java [hunk 3]
```diff
@@ -30,20 +38,40 @@
...
+        ResourceProperties resourceProperties = am.getProperties();
+        try {
+            release = resourceProperties.getInstantProperty(AnnouncementService.RELEASE_DATE).toEpochMilli();
+            date = release;
+        } catch (EntityPropertyTypeException|EntityPropertyNotDefinedException e) { /*No action needed*/ }
```
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/beans/AnnouncementRestBean.java:62]: I'd stick a couple of spaces around the pipe here. Make is it a bit clearer.

**Review by adrianfish — APPROVED** [review]
Just some little changes. Up to you. Does it work okay with the announcements widget/component?

**Comment by stetsche** [comment]
Yes the announcement widget works, I changed the href reference there. Thanks for the reviews :)

**Comment by mpellicer** [comment]
Well done @stetsche and thanks for the review @adrianfish @profmikegreene

**Pull request merged**